### ISSUE TITLE
Potential performance issue with relational reordering

### ISSUE THREAD

**Issue by Cpaczek** [strapi/strapi#15539]
### Required System information

- Node.js version: v18
- Strapi version: 4.5.6
- Database: All
- Operating system: Windows 10

### Describe the bug

https://github.com/strapi/strapi/blob/127c04705751139bb32805e79b6f8619d92b6d34/packages/core/database/lib/entity-manager/index.js#L605-L634

This uses a select to get the most recent order value and adds 1 to it. This could cause a race condition if you try and add a permission order row twice at the same time because it could depend on old max values.

**Comment by derrickmehaffy** [comment]
Engineering, 2 community members/Paul/I were chatting with them about this and this code basically doesn't scale. At some point we will run into a race condition here especially with relational reordering.

We discussed 2 possible solutions but I wanted them to open this ticket in prep for v4.6.0 launch:

- Compound auto-increment value created by the DB
- Implement locking here to prevent race conditions

**Comment by Cpaczek** [comment]
This will have the biggest effect on clustered databases. Currently testing on my cockroach DB branch but it should have a similar effect on other clusters such as maria or MySQL.

```
let arr = []
for (let i = 0; i < 100; i++) {
  arr.push(i.toString())
}
Promise.all(arr.map((i) => strapi.entityService.create("api::blog.blog", {
  data: {
    tester: i,
    comments: [1]
  }
})))
```
If you run this code where blog has a many-to-many relationship to comments and you try and create 100 entries at the same time with a relationship only the last entry will maintain that relationship. 

As you can see here:
![image](https://user-images.githubusercontent.com/25715982/214177295-427b3433-4e0b-4370-af88-9e2747685273.png)

Only the latest entry maintains the relationship.

**Comment by Cpaczek** [comment]
Was able to spin up a maria db multi-master cluster. Maria allows for the rows to be inserted so at first glance it looks okay but if you look at the database you can see there are a whole bunch of duplicate order values.

![image](https://user-images.githubusercontent.com/25715982/214189804-368bb1e1-58b1-40fd-aae4-4abdd1b6bf2a.png)


Also if you try and re order 1 of the relations when there are a lot of them it will take a really long time. For instance, it took 27 seconds to re order 1 of them (Granted there are 8000 relations but still)

![image](https://user-images.githubusercontent.com/25715982/214190405-445f5897-fc6d-4c1d-8a85-84a3ae604269.png)


~~With duplicate order numbers, it causes only the 1st relation to show up in the admin panel:~~

![image](https://user-images.githubusercontent.com/25715982/214191014-8277b324-4d82-4342-a57f-23c6ef801189.png)


Some of them don't even show up. In this screenshot, there should be 200 but only 192 of them get rendered
![image](https://user-images.githubusercontent.com/25715982/214192073-c15eae40-3301-44c1-836b-492dd5c37e84.png)

**Comment by Cpaczek** [comment]
Just found out this isn't a clustering issue we have the same issue on sqlite.

**Comment by Cpaczek** [comment]
This probably also effects the stable version of strapi (4.5.6) so to fix this customers would also need to fix their existing dbs.

**Comment by Cpaczek** [comment]
Also kinda a side note something is converting the column to a float???

![image](https://user-images.githubusercontent.com/25715982/214195226-829c67a3-5cba-4e69-b12e-3716abe1da56.png)

**Comment by Marc-Roig** [comment]
Hello , thank you for raising this problem. We are having dicussions about this.

About the issue, yes , the ordering is calculated on the application level, and not the db. So there could be race conditions like that one.

That could be prevented like @Boegie19 is saying, creating the entities one by one.

Still, this race condition could be an issue and we want to investigate:
- What impact can it have by having duplicated order values.
	- 	Is it critical for usage? If you are creating them in parallel is normal not to have them in order, but you will still see them all. 
	- Maybe it has other implications...
- How we can prevent race conditions like this.

We agree with Derrick about the possible solutions, and we don't see many others than those:

> We discussed 2 possible solutions but I wanted them to open this ticket in prep for v4.6.0 launch:
> 
> Compound auto-increment value created by the DB
> Implement locking here to prevent race conditions

We will use this thread to continue posting about this issue.

**Comment by Marc-Roig** [comment]
And there is a second discussion here, which is the performance degradation of the API.

@Boegie19 for the record, which DB are you using? We plan to replicate your setup and query to see what we can optimize.

FYI: Deleting an entity means reordering all relations that had an order value greater than the one deleted, that will for sure be one of the bottlenecks.

**Comment by Cpaczek** [comment]
> What impact can it have by having duplicated order values.
> * Is it critical for usage? If you are creating them in parallel is normal not to have them in order, but you will still see them all.
> * Maybe it has other implications...

As seen here: https://i.imgur.com/hZXaOqP.gif when you have duplicate order values not all relations will be displayed. 

The proper way would probably compound serial columns but I'm not an expert on that yet. Not sure how reordering would work with that though. 




> @Boegie19 for the record, which DB are you using? We plan to replicate your setup and query to see what we can optimize.

We are using a maria db cluster (master-master) but this bug also happens on sqlite. We following this to set it up: https://blog.devgenius.io/automated-mariadb-replication-using-docker-a585defcc047



> FYI: Deleting an entity means reordering all relations that had an order value greater than the one deleted, that will for sure be one of the bottlenecks.

Even if you delete the last one (i.e last row/ most recent entry) it will take 42 seconds. so that seems to trigger reordering even when it doesn't have to.

**Comment by Marc-Roig** [comment]
> As seen here: [i.imgur.com/hZXaOqP.gif](https://i.imgur.com/hZXaOqP.gif) when you have duplicate order values not all relations will be displayed.

The proper way would probably compound serial columns but I'm not an expert on that yet. Not sure how reordering would work with that though.

It appears there are 100 relations no? reading the example you provided it seems like the right amount.
What I think is happening is that they are not in the order you expect.

> We are using a maria db cluster (master-master) but this bug also happens on sqlite. We following this to set it up: [blog.devgenius.io/automated-mariadb-replication-using-docker-a585defcc047](https://blog.devgenius.io/automated-mariadb-replication-using-docker-a585defcc047)

Alright! Thank you for providing the docs! That will help, we will test with sqlite but also with the maria db cluster just in case.

> Even if you delete the last one (i.e last row/ most recent entry) it will take 42 seconds. so that seems to trigger reordering even when it doesn't have to.
That's very useful info, we will start with that.

**Comment by Cpaczek** [comment]
> It appears there are 100 relations no? reading the example you provided it seems like the right amount.
> What I think is happening is that they are not in the order we would expect (1 to 100).

There are 2 issues with having order numbers overlap. The first being is that entities within groups i.e order # 1 won't have a specified order. But between the groups they are in order.

![image](https://user-images.githubusercontent.com/25715982/214271336-ad89fc7d-171d-4533-b22e-16f257cfcc21.png)


The second being is that in the demo the label says there is 100 relations but if you try and remove all of them (without saving) you can see 8 missings. and when you save, some of them appear.

**Comment by Boegie19** [comment]
@Marc-Roig Just did some more testing and 4.5.4 and above are impacted with the deleting of 1 relationship taking 42 seconds with 10000 relationships https://github.com/strapi/strapi/issues/15472

**Comment by Cpaczek** [comment]
@Marc-Roig I knew there was a reason why the column were set to floats. Instead of re ordering the entire table just use in between values. so if you need to put id:20 between id:5 and id:6 you could do order priority at 5.5 and wouldn't have to re order the table. 

see: https://dba.stackexchange.com/questions/36875/arbitrarily-ordering-records-in-a-table

**Comment by Cpaczek** [comment]
> Alright! Thank you for providing the docs! That will help, we will test with sqlite but also with the maria db cluster just in case.

Probably will happen with a regular maria db database aswell. I don't think the cluster effects much.

**Comment by Boegie19** [comment]
Also the duplicate issue could be solved if you go for the in-between way with just an auto increment and just keep the filed empty when you create it.

**Comment by Cpaczek** [comment]
Ya, it seems like floating point numbers are the best way to do it. If you want to insert between 2 values then just do this
```
let lower_value = 5
let higher_value = 6
let new_in_between_value = ((higher_value - lower_value)/2) + lower_value.
```

You then also need to start the order column by 1 and auto increment the column to prevent the race condition.

**Comment by Marc-Roig** [comment]
> @Marc-Roig Just did some more testing and 4.5.4 and above are impacted with the deleting of 1 relationship taking 42 seconds with 10000 relationships https://github.com/strapi/strapi/issues/15472

After 4.5.0 , we haveen seen some deadlock issues with MySQL and MariaDB. At 4.5.4 we introduced some changes to prevent that and they might have impacted a lot performance. 🙇 That's good info, thank you @Boegie19 

> knew there was a reason why the column were set to floats. Instead of re ordering the entire table just use in between values. so if you need to put id:20 between id:5 and id:6 you could do order priority at 5.5 and wouldn't have to re order the table.

see: [dba.stackexchange.com/questions/36875/arbitrarily-ordering-records-in-a-table](https://dba.stackexchange.com/questions/36875/arbitrarily-ordering-records-in-a-table)

Indeed!! we use the float to reorder it that way. Sorry, forgot to follow on that part.
Thing is, ROW_NUMBER is super useful but is not compatible with mysql 5.7 (which we stilll suport on strapi :( ) . So we decided that for mysql databases we were going to use another logic and we kept the ROW_NUMBER for the others.

**Comment by Marc-Roig** [comment]
> Also the duplicate issue could be solved if you go for the in-between way with just an auto increment and just keep the filed empty when you create it.

What do you mean by `filed empty when you create it?`

> @Marc-Roig Ya, it seems like floating point numbers are the best way to do it. If you want to insert between 2 values then just do this
>
> let lower_value = 5
> let higher_value = 6
> let new_in_between_value = ((higher_value - lower_value)/2) + lower_value
> //5.5
> You then also need to start the order column by 1 and auto increment the column to prevent the race condition.

That could work, but do you think leaving the db with floating numbers could generate some problems in the long run? We could end up having a lot of decimals and getting to apoint where we cannot divide by 2 anymore.


Thank you to both of you for this ideas and the time you are taking. This is very valuable to us and will help us reach the solution faster 🚀

**Comment by Boegie19** [comment]
If you where to make the order auto increment and leave the field empty when you create it the database auto increments the value and adds it to the db so you don't have to do the +1 logic

**Comment by Cpaczek** [comment]
> That could work, but do you think leaving the db with floating numbers could generate some problems in the long run? We could end up having a lot of decimals and getting to apoint where we cannot divide by 2 anymore.
> 
> Thank you to both of you for this ideas and the time you are taking. This is very valuable to us and will help us reach the solution faster 🚀

Yes you would eventually run out. depending on the database you have something around 17 bits of precision (at least that is what it is for cockroachdb not sure about postgres) that means you can divide 1 by 2 about 55 times before you run out of precision. What you could do is only re order when you run out of precision. that means for the vast majority of the time you won't have the performance hit. 



> Indeed!! we use the float to reorder it that way. Sorry, forgot to follow on that part.

Thats how its implemented now? From our testing, there weren't any in-between values when re-ordering relations. it would always re order the entire links database. Are you sure this was implemented?

**Comment by Marc-Roig** [comment]
@Cpaczek indeed I forget to add, we use that logic + then we execute the ROW_NUMBER to transform it to "integers".

Also we are dealing with partial updates. So when we receive the request to add a new relation ( or reorder it), we have to get from the database the order value of the adjacent relations.

That brings some complexities to get the order lower and higher number as in your example:
- We say, add new relation before relation with id 4.  So we can query in the db ID 4 and know the order of the higher value.
- But.. What is the value of the lower if we do not have "integer" like values? 

That last point resulted quite complicated to solve and we decided to not leave order with decimals.

**Comment by Marc-Roig** [comment]
For deleting relationships, we could do what we mentioned somewhere in this thread.
Just not reorder at all, and that should not break anything in theory.

**Comment by Boegie19** [comment]
> But.. What is the value of the lower if we do not have "integer" like values?

https://stackoverflow.com/questions/27086377/get-previous-and-next-row-from-rows-selected-with-where-conditions

you could use someting like this no clue if knex suports it and just ensure it has an unique constrant so that it will fail if it try and do it again since apparently an other instance was added in that space in the main time.

**Comment by Marc-Roig** [comment]
@Boegie19 that is very interesting 👀  didn't know about the lag and lead.
And yes we could have race conditions, (the same as we have now) but I think it could improve the performance.

**Comment by Marc-Roig** [comment]
:( Seems lag and lead functions where adopted in mysql v8  but not before. We do support mysql 5.7 so, it will not be possible to use that.

**Comment by Boegie19** [comment]
> :( Seems lag and lead functions where adopted in mysql v8 but not before. We do support mysql 5.7 so, it will not be possible to use that.

Then it is maybe worth having internally the discussion if you maybe need to drop support for mysql 5.7

The only other solution I see is that you would have to provide both a id of the record before and after it that could fix it

**Comment by Boegie19** [comment]
Can you not just do this for MySQL 5.7:

https://stackoverflow.com/questions/11303532/simulate-lag-function-in-mysql

**Comment by Boegie19** [comment]
@Marc-Roig Sure as soon as you have the quick fix can you link me to the pr so that I can test it?

**Comment by Marc-Roig** [comment]
@Boegie19 yes! :) 
ping @petersg83  as he is working on that too

**Comment by Marc-Roig** [comment]
@Boegie19  https://github.com/strapi/strapi/pull/15554
I think we found a solution that will improve performance by a LOT when using mysql & mariadib.

We found a way to emulate ROW_NUMBER  OVER() using mysql :)

**Comment by Boegie19** [comment]
@Marc-Roig Nice like on strapi's end how much is the improvement?

**Comment by Marc-Roig** [comment]
@Boegie19 With entities with 10k+ entities we were recording 2,3 min for updates. Now it's 500ms :)

**Comment by Boegie19** [comment]
Okay nice one final thing the pr says it fixes this issue while it only fixes the performance issue and not the race condition so this issue should probably stay open until the race condition is also fixed. 

I will wait until @Cpaczek wakes up and then we will test if this fixes all the performance issues aka reordering and deleting. Thanks for the quick PR to fix this. and for the professional and open for decision response to our issue.    👍

**Comment by petersg83** [comment]
In a few minutes the release `0.0.0-a6a9d5c7f90dfc2332d3f97dba9ba9b1f203d0cc` will be out. You'll be able to check the fix with it :)
Should another issue be open to mention the race condition? There is a mix of 2 issues here

**Comment by Boegie19** [comment]
Ye I agree would be better to close this one and make a new one with only the stuff relevant to it. since this would be to much of a read to figure it out. and thanks for making a nightly release to test this ;)

**Comment by Cpaczek** [comment]
Here is separate issue for the race condition: https://github.com/strapi/strapi/issues/15558
